h5py/_hl/dataset.py

```diff
@@ -505,7 +505,7 @@ def __getitem__(self, args):
             # This is necessary because in the case of array types, NumPy
             # discards the array information at the top level.
             new_dtype = readtime_dtype(self.id.dtype, names)
-        mtype = h5t.py_create(new_dtype)
+        mtype = self.id.get_type()
```

Comment:
I suspect this is too simple. It's ignoring what `readtime_dtype` does for the code that was there before, and it isn't affected by `astype`, which the code there before was.

Reply:
I think what `readtime_dtype` does is relevant to `test_compound_vlen_bool`, which is where it's segfaulting in the test run I looked at.

Reply:
Do you have a way around this? I can't seem to find one.

Reply:
I don't fully follow the reasons (see notes below), but I think we need to accept that this code path converts the h5py type to a numpy dtype and back.

If we want to allow reading and writing datasets with a datetime dtype, but not creating those datasets by assigning a numpy array directly, `py_create` (converts dtype to HDF5 TypeID) needs to know the difference between those cases. There's an existing `logical=` flag which seems to be true when creating a dataset and false when reading/writing an existing one, but I'm not sure if it makes sense to reuse that or declare a separate one.

I'd really like more input from other people about how this should work. Maybe this `opaque_dtype` design is more trouble than it's worth. I'll ask on the list.

-----

**Notes from diving into this code**

I don't really understand the comment about 'numpy discards the array information at the top level'. It dates back over 10 years (5a8d34a1a002acddfc1b31c5526a6e3d4de26704), and does not refer to `readtime_dtype`, so presumably it's about why we need to convert the HDF5 type to a numpy dtype and back.

`readtime_dtype` is from only 8 years ago (d8490c808334a5f873982240ac6944ea92e93c0b). The commit message refers to issue 181, but I don't know if that's the same #181 that is now in Github.

`mtype` gets passed to `DatasetID.read()`, and used by HDF5 for type conversion. So it has to be an HDF5 description of the data type of the numpy array we're reading *into*, which may differ from the HDF5 type of the data on disk.

Reply:
I've asked the list for more input: https://groups.google.com/forum/#!topic/h5py/BLIVHygU2GU